sky/utils/controller_utils.py

```diff
@@ -305,7 +330,50 @@ def get_controller_resources(
                     err=f'Expected exactly one resource, got '
                     f'{len(controller_resources)} resources: '
                     f'{controller_resources}'))
-    return list(controller_resources)[0]
+    controller_resources_to_use: resources.Resources = list(
+        controller_resources)[0]
+
+    controller_exist = (global_user_state.get_cluster_from_name(
+        controller.value.name) is not None)
+    if controller_exist or controller_resources_to_use.cloud is not None:
+        return {controller_resources_to_use}
```

Comment:
If we are returning a set of resources anyway, why do we only return the first one but not the whole `controller_resources`, i.e. allowing a user to specify multiple resources for the controller? (If needed, we can add a TODO here)

Also, just wondering, do we have some places depending on the results of the resources returned by this function to decide how many services we can run on the controller? In that case, will changing this to a set cause failure?

Reply:
> why not allowing a user to specify multiple resources for the controller?

That is a good point! I think it is at least feasible to support set of resources (i.e. `any_of`). One thing we need to discuss is what is the behaviour if the controller specified `ordered` resources w/o `cloud` specified and we want to override it with task resources cloud? Consider the following:
```yaml
# ~/.sky/config.yaml
serve:
  controller:
    resources:
      ordered:
        - accelerators: L4
        - accelerators: T4
```
```yaml
# service.yaml
resources:
  any_of:
    - cloud: aws
    - cloud: gcp
```
Should the controller resoruces be a list?
* If it is a list, what is the order of `aws` and `gcp`?
* If it is a set, does it break the semantic of `ordered` in controller resources?

> do we have some places depending on the results of the resources returned by this function to decide how many services we can run on the controller?

No, this is automatically calculated from system memory. Reference here:

https://github.com/skypilot-org/skypilot/blob/48a5c63c42a521513e143f8fd7258e1a9cd30a6f/sky/serve/serve_utils.py#L43-L45

Reply:
Added a TODO first. Thanks!